v2/http-middleware/security.go

```diff
@@ -124,9 +126,20 @@ func handleAccessOrIDToken(ctx context.Context, req *http.Request, header string
 	case jwt.TokenUseID:
 		userID = claims.EnlightUserID
 	case jwt.TokenUseAccess:
+		// Memory map key has been changed to username to provide seamless impersonation.
+		// There is no way to change tokens' claims.Subject in Cognito.
+		// So, we applied a workaround to get the user for impersonation from the access token.
+		// ID token will be generated for the impersonated user in such cases.
+		// In order to be successful, cache key should be set as a username.
+		// If token is created for impersonation, the cache key will be the impersonated user's username.
+		// Required changes for getUserIDByToken for impersonation has been done on users repo.
+		// Since username attribute in Cognito is unique as well, no harm to change the cache key to username.
+		// Ref: https://docs.aws.amazon.com/cognito/latest/developerguide/user-pool-settings-attributes.html
+		username := GetUsernameFromClaims(ctx, claims)
```

Comment:
username is the email right? With this change we are enforcing even harder that changing e-mail is not a feature we will support.

Wouldnt it be better to detect that the token is some kind of "impersonating" token and just avoid caching it? I'm guessing we are thinking most users will not be using impersonating tokens anyway?

Reply:
I don't see a relation between this change and changing email for users feature. This change will only ensures that memory cache is stored with email address as a key instead of Cognito user id (claim.Subject is Cognito user id) If user changes the email, since it won't be found in the cache, middleware will ask SSO service to get the user and then will cache with the user's new email.

We're already detecting access tokens whether it is kind of "impersonating" token. The thing is Cognito doesn't allow to override claims in Access token. I think this is why if the token is ID token then we just return claim.EnlightUserID and we do some magic for Access token to get user's Enlight ID and cache it to reduce load for the SSO api. Otherwise we would add the same claim about Enlight user id to access token. 

Let say **serhat.can@skf.com** is admin user and logged in Enlight with own credentials and doing somethings for himself. There is an access token with its subject is "1234" . Suddenly he got a request to solve an issue on behalf of a customer whose email is **serhat.can@hiq.se** . He creates tokens for impersonation. He gets an access token with an addtional indication that can be recognized an impersonated token but since subject of the token is not changeble, it will get the impersonation token with same sub "1234" . As you can see it will be a problem since there will be a record in the map  for the key "1234" which maps it to admin EnlightID. It should point to the impersonated user's ID.

We can add user's EnlightID as we add impersonation indication to the Access token as well. There is only one workaround for Cognito tokens that "cognito:groups" attribute is mutal between ID and Access tokens. So we add a string with a prefix to distinguish impersonated token into that area. As an example of impersonated access token is attached. In this way, we can avoid caching not only for impersonating but also for all.

![image](https://user-images.githubusercontent.com/105059106/171820488-58e2124b-aed0-4773-8c78-9d1ac68cbfa1.png)

Reply:
@mensurm fyi

Reply:
this is not related anymore since we changed  the structure of the tokens after this comment.